pkg/apis/workflow/v1alpha1/event_types.go

```diff
@@ -40,6 +40,17 @@ type Submit struct {
 	// WorkflowTemplateRef the workflow template to submit
 	WorkflowTemplateRef WorkflowTemplateRef `json:"workflowTemplateRef" protobuf:"bytes,1,opt,name=workflowTemplateRef"`
 
+	// WorkflowName optional means to customize the name of the submitted workflow
+	WorkflowName *WorkflowName `json:"workflowName,omitempty" protobuf:"bytes,2,opt,name=workflowName"`
```

Comment:
this is great. 

* We could allow for future changes, e.g. templating labels and annotations.
* We could make the names fully based on the expression environment - no suffix. 

```yaml
metadata:
  name: payload.message + timestamp
```

Useful as follows:

```yaml
metadata:
  name: payload.github.repo + "-" + payload.github.sha1
```

Maybe in the future:

```yaml
metadata:
  annotations: 
    sha1: payload.github.sha1
```